source/common/stats/thread_local_store.cc

```diff
@@ -86,6 +86,14 @@ void ThreadLocalStoreImpl::releaseScopeCrossThread(ScopeImpl* scope) {
   }
 }
 
+std::string ThreadLocalStoreImpl::getTagsForName(const std::string& name, std::vector<Tag>& tags) {
+  std::string tag_extracted_name = name;
+  for (const TagExtractor& tag_extractor : tag_extractors_) {
+    tag_extracted_name = tag_extractor.updateTags(tag_extracted_name, tags);
```

Comment:
I'm a bit confused by this.. I would have naively thought we scan along the list until we find a match (either forward or backward), then stop with the tag extracted name and tag set we want to use. The loop here is a bit different, it seems to allow interaction between multiple tag rules, chaining. Can you elaborate on what the rationale is?

Reply:
I think I understand what you're getting at, and I think that's a viable way of doing it, but the current design allows tags to be somewhat independent of one another, meaning each regex represents exactly one tag.  If I were to add a new tag that would come later in the string after `cluster.<cluster_name>.` in the existing setup, I don't need to capture the cluster name as a part of that regex.  Let's assume I want to capture every string that ends with `_rq_\d{3}` as a tag, which is similar to one of the existing tags.  With your suggestion, I would need to write a different regex for each of the possible strings that precedes that (or make a single regex that captures all possibilities).  It's obviously not combinatorial because most tags only show up in one place, but it seems less flexible overall, especially if you're trying to capture a common request qualifier (like response code) that comes at the end of a few differently prefixed metrics.  I also think this would lead to a more complex regex grammar (more capture groups, possibly named).  However, I do agree that the current approach can lead to unexpected ordering interactions.  The existing tests ensure that the defaults work as expected both in forward iteration and back, and any custom additions have a defined ordering according to the config so that users can depend on a specified ordering.

Reply:
Yeah, OK, this makes sense. You might want to comment on this somewhere if not already.